File: tests/cpp/test_cobertura.cpp
/**
 * @file test_cobertura.cpp
 * @brief Tests del análisis de déficit/cobertura (DS 024, Art. 252 f/g).
 */

#include <gtest/gtest.h>
#include <stdexcept>

#include "ventpy/cobertura.hpp"

using namespace ventpy;

// ============================================================================
// compare_zone — medición directa (nivel puro)
// ============================================================================

TEST(CoverageZone, DirectaCumpleSinAdvertencias) {
    ZoneMeasurement m;
    m.zone_name = "Rampa 4200";
    m.q_measured_m3min = 250.0;

    auto r = CoverageCalculator::compare_zone(207.0, m);

    EXPECT_EQ(r.zone_name, "Rampa 4200");
    EXPECT_DOUBLE_EQ(r.q_required_m3min, 207.0);
    EXPECT_DOUBLE_EQ(r.q_measured_m3min, 250.0);
    EXPECT_NEAR(r.coverage_ratio, 250.0 / 207.0, 1e-9);   // 1.2077 crudo
    EXPECT_TRUE(r.compliant);
    EXPECT_FALSE(r.near_deficit_warning);                 // 1.2077 > 1.10
    EXPECT_FALSE(r.overventilated);                       // 1.2077 < 1.5
    EXPECT_DOUBLE_EQ(r.deficit_m3min, 0.0);
    EXPECT_NE(r.regulation_ref.find("252"), std::string::npos);
}

TEST(CoverageZone, DeficitUsaSafetyCeil) {
    ZoneMeasurement m;
    m.zone_name = "Frente N-02";
    m.q_measured_m3min = 90.2;

    auto r = CoverageCalculator::compare_zone(100.0, m);

    EXPECT_FALSE(r.compliant);
    // deficit = ceil(100 - 90.2) = ceil(9.8) = 10 — NUNCA 9
    EXPECT_DOUBLE_EQ(r.deficit_m3min, 10.0);
    EXPECT_NEAR(r.coverage_ratio, 0.902, 1e-9);
}

TEST(CoverageZone, CoberturaJustaGeneraAdvertenciaDeMargen) {
    ZoneMeasurement m;
    m.zone_name = "Tajeo 380";
    m.q_measured_m3min = 500.0;

    auto r = CoverageCalculator::compare_zone(500.0, m);

    EXPECT_TRUE(r.compliant);            // medido >= requerido
    EXPECT_TRUE(r.near_deficit_warning); // 1.00 < 1.10 (margen ingenieril 10%)
    EXPECT_DOUBLE_EQ(r.deficit_m3min, 0.0);
}

TEST(CoverageZone, SobreVentilacionAdvierte) {
    ZoneMeasurement m;
    m.zone_name = "Bypass 12";
    m.q_measured_m3min = 160.0;

    auto r = CoverageCalculator::compare_zone(100.0, m);

    EXPECT_TRUE(r.compliant);
    EXPECT_TRUE(r.overventilated);       // 1.6 > 1.5 (factor ingenieril)
}

TEST(CoverageZone, UmbralesPersonalizados) {
    CoverageParams p;
    p.warning_margin = 0.30;             // exigir 130% para no advertir
    p.overventilation_factor = 2.0;

    ZoneMeasurement m;
    m.zone_name = "Z";
    m.q_measured_m3min = 120.0;

    auto r = CoverageCalculator::compare_zone(100.0, m, p);
    EXPECT_TRUE(r.near_deficit_warning);  // 1.2 < 1.3
    EXPECT_FALSE(r.overventilated);       // 1.2 < 2.0
}

// ============================================================================
// compare_zone — validación en frontera
// ============================================================================

TEST(CoverageZone, FuenteAmbiguaLanza) {
    ZoneMeasurement m;
    m.zone_name = "Z";
    m.q_measured_m3min = 100.0;
    m.stations.push_back({"E-1", 10.0, 1.0});   // ambas fuentes

    EXPECT_THROW(CoverageCalculator::compare_zone(100.0, m),
                 std::invalid_argument);
}

TEST(CoverageZone, SinFuenteLanza) {
    ZoneMeasurement m;
    m.zone_name = "Z";

    EXPECT_THROW(CoverageCalculator::compare_zone(100.0, m),
                 std::invalid_argument);
}

TEST(CoverageZone, RequeridoNoPositivoLanza) {
    ZoneMeasurement m;
    m.zone_name = "Z";
    m.q_measured_m3min = 100.0;

    EXPECT_THROW(CoverageCalculator::compare_zone(0.0, m),
                 std::invalid_argument);
}

TEST(CoverageZone, MedidoNegativoLanza) {
    ZoneMeasurement m;
    m.zone_name = "Z";
    m.q_measured_m3min = -5.0;

    EXPECT_THROW(CoverageCalculator::compare_zone(100.0, m),
                 std::invalid_argument);
}

TEST(CoverageZone, ParamsInvalidosLanzan) {
    ZoneMeasurement m;
    m.zone_name = "Z";
    m.q_measured_m3min = 100.0;

    CoverageParams margen_negativo;
    margen_negativo.warning_margin = -0.1;
    EXPECT_THROW(CoverageCalculator::compare_zone(100.0, m, margen_negativo),
                 std::invalid_argument);

    CoverageParams factor_bajo;
    factor_bajo.overventilation_factor = 1.0;   // debe ser > 1
    EXPECT_THROW(CoverageCalculator::compare_zone(100.0, m, factor_bajo),
                 std::invalid_argument);

    CoverageParams velocidades_invertidas;
    velocidades_invertidas.min_velocity_mpm = 300.0;   // min >= max
    EXPECT_THROW(CoverageCalculator::compare_zone(100.0, m, velocidades_invertidas),
                 std::invalid_argument);
}

// ============================================================================
// compare_zone — estaciones de aforo + velocidad Art. 248
// ============================================================================

TEST(CoverageStations, SumaEstacionesParalelas) {
    ZoneMeasurement m;
    m.zone_name = "Nivel 380";
    m.stations.push_back({"E-1", 10.0, 1.0});   // Q = 10 × 1.0 × 60 = 600
    m.stations.push_back({"E-2", 5.0, 0.5});    // Q = 5 × 0.5 × 60 = 150

    auto r = CoverageCalculator::compare_zone(700.0, m);

    ASSERT_EQ(r.stations.size(), 2u);
    EXPECT_DOUBLE_EQ(r.stations[0].q_station_m3min, 600.0);
    EXPECT_DOUBLE_EQ(r.stations[0].velocity_mpm, 60.0);
    EXPECT_TRUE(r.stations[0].velocity_ok);       // 20 <= 60 <= 250
    EXPECT_DOUBLE_EQ(r.stations[1].q_station_m3min, 150.0);
    EXPECT_DOUBLE_EQ(r.q_measured_m3min, 750.0);  // suma paralela
    EXPECT_TRUE(r.compliant);                     // 750 >= 700
}

TEST(CoverageStations, VelocidadBajoMinimoAdvierte) {
    ZoneMeasurement m;
    m.zone_name = "Z";
    m.stations.push_back({"E-1", 12.0, 0.30});  // 18 m/min < 20 (Art. 248)

    auto r = CoverageCalculator::compare_zone(100.0, m);

    ASSERT_EQ(r.stations.size(), 1u);
    EXPECT_FALSE(r.stations[0].velocity_ok);
    EXPECT_NE(r.stations[0].warning.find("248"), std::string::npos);
    // El caudal SÍ se contabiliza aunque la velocidad esté fuera de rango
    EXPECT_DOUBLE_EQ(r.q_measured_m3min, 216.0);
}

TEST(CoverageStations, VelocidadSobreMaximoAdvierte) {
    ZoneMeasurement m;
    m.zone_name = "Z";
    m.stations.push_back({"E-1", 4.0, 4.5});    // 270 m/min > 250

    auto r = CoverageCalculator::compare_zone(100.0, m);
    EXPECT_FALSE(r.stations[0].velocity_ok);
    EXPECT_NE(r.stations[0].warning.find("248"), std::string::npos);
}

TEST(CoverageStations, AnfoElevaMinimoA25) {
    ZoneMeasurement m;
    m.zone_name = "Z";
    m.stations.push_back({"E-1", 10.0, 0.35});  // 21 m/min

    // Sin ANFO: 21 >= 20 → ok
    auto sin_anfo = CoverageCalculator::compare_zone(100.0, m);
    EXPECT_TRUE(sin_anfo.stations[0].velocity_ok);

    // Con ANFO: mínimo efectivo 25 (Art. 248) → 21 < 25 advierte
    CoverageParams p;
    p.anfo_or_blasting_agents = true;
    auto con_anfo = CoverageCalculator::compare_zone(100.0, m, p);
    EXPECT_FALSE(con_anfo.stations[0].velocity_ok);
    EXPECT_NE(con_anfo.stations[0].warning.find("ANFO"), std::string::npos);
}

TEST(CoverageStations, VelocidadCeroEsMedicionValida) {
    // Labor sin flujo: dato real de levantamiento — Q = 0, advertencia por
    // velocidad, sin excepción.
    ZoneMeasurement m;
    m.zone_name = "Z";
    m.stations.push_back({"E-1", 8.0, 0.0});

    auto r = CoverageCalculator::compare_zone(100.0, m);
    EXPECT_DOUBLE_EQ(r.q_measured_m3min, 0.0);
    EXPECT_FALSE(r.compliant);
    EXPECT_FALSE(r.stations[0].velocity_ok);
}

TEST(CoverageStations, AreaNoPositivaLanza) {
    ZoneMeasurement m;
    m.zone_name = "Z";
    m.stations.push_back({"E-1", 0.0, 1.0});
    EXPECT_THROW(CoverageCalculator::compare_zone(100.0, m),
                 std::invalid_argument);
}

TEST(CoverageStations, VelocidadNegativaLanza) {
    ZoneMeasurement m;
    m.zone_name = "Z";
    m.stations.push_back({"E-1", 10.0, -0.5});
    EXPECT_THROW(CoverageCalculator::compare_zone(100.0, m),
                 std::invalid_argument);
}

File: tests/python/test_cobertura.py
"""
Tests de integracion Python del analisis de cobertura/deficit.

Verifica que los bindings nanobind exponen correctamente CoverageCalculator
(compare_zone / analyze_survey) y las structs de entrada/resultado.

Normativa: DS 024-2016-EM (mod. DS 023-2017-EM), Art. 252 lit. f/g.
Numeros tomados del probe SP-2 / tests/cpp/test_cobertura.cpp (ya validados).
"""

import math

import pytest

import ventpy


def make_zone(name, altitude, measured_direct):
    """Espejo de make_zone() en tests/cpp/test_cobertura.cpp."""
    z = ventpy.ZoneSurvey()
    z.zone_name = name
    z.input = ventpy.VentilationInput()
    z.input.zone_type = ventpy.ZoneType.DevelopmentFace
    z.input.num_workers = 10
    z.input.altitude_masl = altitude
    z.measurement = ventpy.ZoneMeasurement()
    z.measurement.zone_name = name
    z.measurement.q_measured_m3min = measured_direct
    return z


# ============================================================================
# compare_zone - medicion directa
# ============================================================================


class TestCompareZoneDirect:
    def test_compliant_reports_ratio(self):
        m = ventpy.ZoneMeasurement()
        m.zone_name = "Rampa 4200"
        m.q_measured_m3min = 250.0

        r = ventpy.CoverageCalculator.compare_zone(207.0, m)

        assert r.zone_name == "Rampa 4200"
        assert r.q_required_m3min == 207.0
        assert r.q_measured_m3min == 250.0
        assert r.coverage_ratio == pytest.approx(250.0 / 207.0)
        assert r.compliant is True
        assert r.deficit_m3min == 0.0
        assert "252" in r.regulation_ref

    def test_deficit_uses_safety_ceil(self):
        # deficit = ceil(100 - 90.2) = ceil(9.8) = 10 -- NUNCA 9
        m = ventpy.ZoneMeasurement()
        m.zone_name = "Frente N-02"
        m.q_measured_m3min = 90.2

        r = ventpy.CoverageCalculator.compare_zone(100.0, m)

        assert r.compliant is False
        assert r.deficit_m3min == 10.0
        assert r.coverage_ratio == pytest.approx(0.902)

    def test_measured_infinite_raises(self):
        # FIX 1 (pre-bindings): inf pasaba require_non_negative (inf >= 0)
        # y reportaba compliant=True de forma espuria. Debe lanzar.
        m = ventpy.ZoneMeasurement()
        m.zone_name = "Z"
        m.q_measured_m3min = math.inf

        with pytest.raises(ValueError):
            ventpy.CoverageCalculator.compare_zone(100.0, m)


# ============================================================================
# compare_zone - estaciones de aforo
# ============================================================================


class TestCompareZoneStations:
    def test_stations_sum_in_parallel(self):
        m = ventpy.ZoneMeasurement()
        m.zone_name = "Nivel 380"
        e1 = ventpy.AirflowStation()
        e1.station_id = "E-1"
        e1.area_m2 = 10.0
        e1.velocity_mps = 1.0  # Q = 10 x 1.0 x 60 = 600
        e2 = ventpy.AirflowStation()
        e2.station_id = "E-2"
        e2.area_m2 = 5.0
        e2.velocity_mps = 0.5  # Q = 5 x 0.5 x 60 = 150
        m.stations = [e1, e2]

        r = ventpy.CoverageCalculator.compare_zone(700.0, m)

        assert len(r.stations) == 2
        assert r.stations[0].q_station_m3min == 600.0
        assert r.stations[1].q_station_m3min == 150.0
        assert r.q_measured_m3min == 750.0
        assert r.compliant is True

    def test_reasignar_medicion_a_estaciones(self):
        # Reasignar la fuente de medicion (de directa a estaciones) debe
        # dejar la anterior sin efecto: compare_zone usa exactamente UNA
        # fuente (has_direct == has_stations lanza), la vigente al momento
        # de la llamada.
        m = ventpy.ZoneMeasurement()
        m.zone_name = "Z-reasignada"
        m.q_measured_m3min = 100.0

        m.q_measured_m3min = None
        e1 = ventpy.AirflowStation()
        e1.station_id = "E-1"
        e1.area_m2 = 10.0
        e1.velocity_mps = 1.0  # Q = 10 x 1.0 x 60 = 600
        m.stations = [e1]

        r = ventpy.CoverageCalculator.compare_zone(500.0, m)

        assert len(r.stations) == 1
        assert r.q_measured_m3min == 600.0
        assert r.compliant is True

    def test_velocity_out_of_range_warns_248(self):
        m = ventpy.ZoneMeasurement()
        m.zone_name = "Z"
        e1 = ventpy.AirflowStation()
        e1.station_id = "E-1"
        e1.area_m2 = 12.0
        e1.velocity_mps = 0.30  # 18 m/min < 20 (Art. 248)
        m.stations = [e1]

        r = ventpy.CoverageCalculator.compare_zone(100.0, m)

        assert len(r.stations) == 1
        assert r.stations[0].velocity_ok is False
        assert "248" in r.stations[0].warning
        # El caudal SI se contabiliza aunque la velocidad este fuera de rango
        assert r.q_measured_m3min == pytest.approx(216.0)


# ============================================================================
# analyze_survey - balance de mina via Governor (E2E)
# ============================================================================


class TestAnalyzeSurvey:
    def test_global_covers_with_zone_in_deficit(self):
        zones = [
            make_zone("Rampa 4200", 4200.0, 270.0),
            make_zone("Frente N-02", 2500.0, 150.0),
        ]

        r = ventpy.CoverageCalculator.analyze_survey(
            zones, ventpy.RegulatoryConfig.peru())

        assert len(r.zones) == 2
        assert r.zones[0].q_required_m3min == 207.0
        assert r.zones[1].q_required_m3min == 207.0
        assert r.zones[0].compliant is True     # 270 >= 207
        assert r.zones[1].compliant is False    # 150 < 207
        assert r.zones[1].deficit_m3min == 57.0

        assert r.q_required_total_m3min == 414.0
        assert r.q_measured_total_m3min == 420.0
        assert r.global_compliant is True       # Art. 252.f: Sum(med) >= Sum(req)
        assert r.all_zones_compliant is False   # Art. 252.g: zona en deficit
        assert r.compliant is False             # criterio estricto: ambos
        assert r.deficit_total_m3min == 0.0

        # Cada zona lleva su desglose completo del Governor
        assert r.zones[0].demand is not None
        assert r.zones[0].demand.standard == ventpy.RegulatoryStandard.DS024_Peru

        deficit_warned = any("Frente N-02" in w for w in r.warnings)
        assert deficit_warned

    def test_general_mine_zone_raises(self):
        z = make_zone("Mina total", 2500.0, 1000.0)
        z.input.zone_type = ventpy.ZoneType.GeneralMine  # doble conteo -> prohibido

        with pytest.raises(ValueError):
            ventpy.CoverageCalculator.analyze_survey(
                [z], ventpy.RegulatoryConfig.peru())
